src/Entity/EntityId.php

```diff
@@ -30,6 +30,37 @@ public function getSerialization() {
 	}
 
 	/**
+	 * Returns the foreign repository name or empty string for local entities.
+	 * @return string
+	 */
+	public function getRepoName() {
+		if ( $this->isForeign() ) {
+			return strtok( $this->serialization, ':' );
+		}
+
+		return '';
+	}
+
+	/**
+	 * @return bool
+	 */
+	public function isForeign() {
+		return strpos( $this->serialization, ':' ) > 0;
+	}
+
+	protected function sanitizeIdSerialization( $id ) {
+		return $this->upcaseLocalId( ltrim( $id, ':' ) );
+
+	}
+
+	private function upcaseLocalId( $id ) {
+		$parts = explode( ':', $id );
+		$parts[count( $parts ) - 1] = strtoupper( end( $parts ) );
```

Comment:
Local IDs never have colons in them (we decided that), but foreign IDs can be "chained": foo:bar:Q123 is Q123 in the repo that is called bar by the repo foo. That's also how interwiki prefixes work. You can do en:fr:wiktionary:de:wikibooks:it:wikipedia:en:Foo. This will result in a series of HTTP redirects, eventually giving you the Foo page on English Wikipedia.

My personal preference for code re-use would be: protected static upcaseLocalId() called from the constructors of ItemId and ProeprtyId.